lib/spack/spack/fetch_strategy.py

```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+        # with git as well.
+        if not spack.config.get('config:verify_ssl'):
+            git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
 
-        return self._git
+        return git
 
     @property
     def cachable(self):
-        return self.cache_enabled and bool(self.commit or self.tag)
+        return (self.cache_enabled and
+                not self.ref.always_check_for_updates_but_accept_airgapped())
 
     def source_id(self):
-        return self.commit or self.tag
+        return self.ref.refspec()
 
     def mirror_id(self):
-        repo_ref = self.commit or self.tag or self.branch
-        if repo_ref:
-            repo_path = url_util.parse(self.url).path
-            result = os.path.sep.join(['git', repo_path, repo_ref])
-            return result
+        repo_path = url_util.parse(self.url).path
+        return os.path.sep.join(['git', repo_path, self.ref.refspec()])
 
     def _repo_info(self):
-        args = ''
+        return '{0} {1}'.format(self.url, self.ref.repo_info_for_reference())
+
+    @memoized
+    def _cache_path(self):
+        # type: () -> str
+        cache_path = spack.caches.fetch_cache.persistent_cache_dir_for(self)
+        if self._initialize_git_repo(cache_path):
+            tty.debug('Initialized git repository: {0}'.format(self._repo_info()))
+        else:
+            tty.debug('Already fetched {0}'.format(cache_path))
+        return cache_path
+
+    def _dormant_branch_name(self, cache_path):
+        # type: (str) -> str
+        if self.branch == self._special_dormant_branch_name:
+            raise FetcherConflict(dedent("""\
+            The requested branch {0} is the same name that Spack uses internally for its
+            git repo caching strategy. Using this branch name may cause errors.
+            The canonical git repo for this remote is located at {1}.
+            """).format(self.branch, cache_path))
+        return self._special_dormant_branch_name
+
+    def _is_on_dormant_branch(self, cache_path):
+        # type: (str) -> bool
+        message = 'On branch {0}'.format(self._dormant_branch_name(cache_path))
+        try:
+            with working_dir(cache_path):
+                return self.git('status', output=str, error=str).startswith(message)
+        except ProcessError:
+            return False
+
+    def _is_initialized_git_repo(self, cache_path):
+        # type: (str) -> bool
+        git_dir = os.path.join(cache_path, '.git')
+        return os.path.isdir(git_dir) and self._is_on_dormant_branch(cache_path)
+
+    def _initialize_git_repo(self, cache_path):
+        # type: (str) -> bool
+        """This method returns True when it was force to reinitialize the cache repo."""
+        if self._is_initialized_git_repo(cache_path):
+            return False
+        # If this is the first time we've tried checking out this repo, we'll need to
+        # create this directory.
+        mkdirp(cache_path)
```

Comment:
This will cause a problem if the directory already existed and was on a different branch.

Reply:
If the directory exists, I think we should check whether it is a git repo. If not, we should error out.

Then we should check whether it points to the same remote we want. If not, we should error out.

Then we can just checkout the dormant branch. This will eliminate the failure mode in which we end up on the wrong branch somehow, or a  user plays around looking at the different refs in the cache. I believe the only performance loss is in cases that we would error on the existing code, so I'm not too worried about the cost of these checks.

Is there something I'm missing in this?

Reply:
> This will cause a problem if the directory already existed and was on a different branch. Is there a failure mode in which that is possible?

The `cache_path` is only ever set to the `__dormant` branch. If the `cache_path` is already initialized, then it is on the correct branch. We do not ever modify the checked-out `HEAD` of the cached git repo after creating it -- we will only fetch refs, which are then visible to every *worktree*. Worktrees are not at all considered nor do they matter in this method -- we strictly handle worktrees in the actual `.fetch()` method.

Worktrees will only ever be pointing to a single type of (immutable) ref, so the only time we ever need to worry about what's been in any type of git repo directory beforehand is in the case of *branches*. This is the reason I elected to use the extremely long name `.always_check_for_updates_but_accept_airgapped()`, because that conditional logic was the only case in which we have to consider the problem of mutable filesystem state. I think that the current name of that method isn't as helpful as it could be, but I *do* think that it is a good idea to make that conditional check into a verbose method name so people don't start relying on it elsewhere. "Whether the ref type is mutable" is really, really an implementation detail, imo, so I don't want to convert that judgement into a more generic predicate method.

Reply:
I was able to work around this by generating a UUID!